xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,444 @@
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw (exception instanceof IOException) ? (IOException) exception :
+              new IOException(exception);
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      StatusException statusException = Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException();
+      listener.onNotServing(statusException);
+      initialStartFuture.set(statusException);
+      return;
+    }
+    xdsClient = xdsClientPool.getObject();
+    // TODO(chengyuanzhang): add an API on XdsClient indicating if it is using v3, don't get
+    //  from bootstrap.
+    boolean useProtocolV3 = xdsClient.getBootstrapInfo().getServers().get(0).isUseProtocolV3();
+    String listenerTemplate = xdsClient.getBootstrapInfo().getServerListenerResourceNameTemplate();
+    if (!useProtocolV3 || listenerTemplate == null) {
+      StatusException statusException =
+          Status.UNAVAILABLE.withDescription(
+              "Can only support xDS v3 with listener resource name template").asException();
+      listener.onNotServing(statusException);
+      initialStartFuture.set(statusException);
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+      return;
+    }
+    discoveryState = new DiscoveryState(listenerTemplate.replaceAll("%s", listenerAddress));
+  }
+
+  @Override
+  public Server shutdown() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (!delegate.isShutdown()) {
+          delegate.shutdown();
+        }
+        internalShutdown();
+      }
+    });
+    return this;
+  }
+
+  @Override
+  public Server shutdownNow() {
+    if (!shutdown.compareAndSet(false, true)) {
+      return this;
+    }
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (!delegate.isShutdown()) {
+          delegate.shutdownNow();
+        }
+        internalShutdown();
+      }
+    });
+    return this;
+  }
+
+  // Must run in SynchronizationContext
+  private void internalShutdown() {
+    logger.log(Level.FINER, "Shutting down XdsServerWrapper");
+    if (discoveryState != null) {
+      discoveryState.shutdown();
+    }
+    if (xdsClient != null) {
+      xdsClient = xdsClientPool.returnObject(xdsClient);
+    }
+    if (sharedTimeService) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
+    }
+    if (restartTimer != null) {
+      restartTimer.cancel();
+    }
+    isServing = false;
+    internalTerminationLatch.countDown();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown.get();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return internalTerminationLatch.getCount() == 0 && delegate.isTerminated();
```

Comment:
nit: `internalTerminationLatch.await(0L, TimeUnit.MILLISECONDS)` would be the recommended way since getCount() "is typically used for debugging and testing purposes."